rust/core-lib/src/movement.rs

```diff
@@ -88,14 +84,84 @@ fn vertical_motion(
     } else {
         line.saturating_add(line_delta as usize)
     };
-    let n_lines = view.line_of_offset(text, text.len());
     if line > n_lines {
         return (text.len(), Some(col));
     }
     let new_offset = view.line_col_to_offset(text, line, col);
     (new_offset, Some(col))
 }
 
+/// Compute movement based on vertical motion by the given number of lines skipping
+/// any line that is shorter than the current cursor position.
+fn vertical_motion_exact_pos(
+    r: SelRegion,
+    view: &View,
+    text: &Rope,
+    line_delta: isize,
+    modify: bool,
+) -> (usize, Option<HorizPos>) {
+    let (col, init_line, n_lines) = selection_position(r, view, text, line_delta, modify);
+
+    let mut line_length = view.offset_of_line(text, init_line.saturating_add(1))
+        - view.offset_of_line(text, init_line);
+    if line_delta < 0 && (-line_delta as usize) > init_line {
+        return (view.line_col_to_offset(text, init_line, col), Some(col));
+    }
+    let mut line = if line_delta < 0 {
+        init_line - (-line_delta as usize)
+    } else {
+        init_line.saturating_add(line_delta as usize)
+    };
+
+    // If the active columns is longer than the current line, use the current line length.
+    let col = if line_length < col { line_length - 1 } else { col };
+
+    loop {
+        line_length =
+            view.offset_of_line(text, line.saturating_add(1)) - view.offset_of_line(text, line);
+
+        // If the line is longer than the current cursor position, break.
+        // We use > instead of >= because line_length includes newline.
+        if line_length > col {
```

Comment:
I also might consider combining these two statements into a single compound statement, eg, `if x || y || z { break; }`, although this is mostly a style/personal preference thing, so not a blocker.

Reply:
Well in the case of going past the line limit (first or last) I want to set the line to the init_line so I think this makes more sense as two if statements. Unless I am missing something I am going to leave as is.